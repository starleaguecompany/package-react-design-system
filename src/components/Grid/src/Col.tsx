import * as React from 'react';
import { useStyles } from '@starleaguecompany/package-react-utils';

import { ColProps } from '../types/Col.types';
import RowContext from './RowContext';
import styles from '../styles/Col.module.less';

/**
 * @description Column element for Grid system.
 *
 * @component
 * @example
 * ```jsx
 * <Col span={3}>
 *   Column
 * </Button>
 * ```
 */
const Col = React.forwardRef<HTMLDivElement, ColProps>((props, ref) => {
  const { className, span, order, breakpoints, style, children, ...restProps } = props;
  const cx = useStyles(styles);
  // const [match] = useMediaQuery()

  // const breakPointProps = React.useMemo(() => {
  //   if (breakpoints && match) {
  //     // @ts-ignore
  //     return breakpoints[match] || {}
  //   }
  //
  //   return {}
  // }, [match, breakpoints])

  const classNames = cx(className, 'container', {
    // [`span-${breakPointProps.span || span}`]: breakPointProps.span || span,
    [`span-${span}`]: span,
  });
  const { gutter } = React.useContext(RowContext);
  const mergedStyle: React.CSSProperties = {
    ...style,
    // order: breakPointProps.order || order || undefined,
    order: order || undefined,
    // ...breakPointProps.style,
  };

  // Horizontal gutter use padding
  if (gutter && gutter[0] > 0) {
    const horizontalGutter = gutter[0] / 2;
    mergedStyle.paddingLeft = horizontalGutter;
    mergedStyle.paddingRight = horizontalGutter;
  }

  // Vertical gutter use padding when gap not support
  if (gutter && gutter[1] > 0) {
    const verticalGutter = gutter[1] / 2;
    mergedStyle.paddingTop = verticalGutter;
    mergedStyle.paddingBottom = verticalGutter;
  }

  return (
    <div data-qa="Col" ref={ref} className={classNames} style={{ ...mergedStyle }} {...restProps}>
      {children}
    </div>
  );
});

Col.defaultProps = {};

export default Col;
