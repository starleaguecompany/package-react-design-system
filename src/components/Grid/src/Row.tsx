import * as React from 'react'
import { useStyles } from '@starleaguecompany/package-react-utils'

import { RowProps } from '../types/Row.types'
import RowContext from './RowContext'
import styles from '../styles/Row.module.less'

/**
 * @description Container element for Grid system.
 *
 * @component
 * @example
 * ```jsx
 * <Row gutter={[20, 20]} wrap>
 *   ...columns
 * </Button>
 * ```
 */
const Row = React.forwardRef<HTMLDivElement, RowProps>((props, ref) => {
  const { className, gutter, align = 'top', justify, wrap: wrap, style, children, ...restProps } = props
  const cx = useStyles(styles)

  const classNames = cx(className, 'container', {
    wrap: wrap,
    [`align-${align}`]: align,
    [`justify-${justify}`]: justify,
  })

  const gutters = React.useMemo(() => {
    const results: [number, number] = [0, 0]
    const normalizedGutter = Array.isArray(gutter) ? gutter : [gutter, 0]

    normalizedGutter.forEach((g, index) => {
      results[index] = g || 0
    })

    return results
  }, [gutter])

  const rowStyle: React.CSSProperties = { ...style }
  const horizontalGutter = gutters[0] > 0 ? gutters[0] / -2 : undefined

  if (horizontalGutter) {
    rowStyle.marginLeft = horizontalGutter
    rowStyle.marginRight = horizontalGutter
  }

  const rowContext = React.useMemo(() => ({ gutter: gutters, wrap }), [gutters, wrap])

  return (
    <RowContext.Provider value={rowContext}>
      <div data-qa="Row" ref={ref} className={classNames} style={{ ...rowStyle }} {...restProps}>
        {children}
      </div>
    </RowContext.Provider>
  )
})

Row.defaultProps = {
  gutter: 0,
  align: 'top',
  justify: 'start',
  wrap: true,
}

export default Row
