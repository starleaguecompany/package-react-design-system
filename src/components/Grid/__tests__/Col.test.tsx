import * as React from 'react'
import { render, screen } from '@testing-library/react'

import Row from '../src/Row'
import Col from '../src/Col'

describe('Col', () => {
  test('should render correctly with no props', () => {
    render(<Col />)

    expect(screen.getByTestId('Col')).toMatchSnapshot()
    expect(screen.getByTestId('Col')).toHaveClass('container', { exact: true })
  })

  test('should render correctly with attributes', () => {
    render(
      <Col id="test-id" className="test-class">
        Testing col
      </Col>
    )

    expect(screen.getByTestId('Col')).toMatchSnapshot()
    expect(screen.getByTestId('Col')).toHaveClass('test-class')
    expect(screen.getByTestId('Col')).toHaveAttribute('id', 'test-id')
  })

  test('should render correctly with text', () => {
    render(<Col>Testing col</Col>)

    expect(screen.getByTestId('Col')).toMatchSnapshot()
    expect(screen.getByTestId('Col')).toHaveTextContent('Testing col')
    expect(screen.getByText('Testing col')).toBeVisible()
  })

  test('should render correctly with span prop', () => {
    render(<Col span={6}>Testing col</Col>)

    expect(screen.getByTestId('Col')).toMatchSnapshot()
    expect(screen.getByTestId('Col')).toHaveClass('span-6')
  })

  test('should render correctly with order prop', () => {
    render(
      <Row>
        <Col order={3}>Testing col 1</Col>
        <Col order={1}>Testing col 2</Col>
        <Col order={2}>Testing col 3</Col>
      </Row>
    )

    expect(screen.getByTestId('Row')).toMatchSnapshot()
    expect(screen.getByText('Testing col 1')).toHaveStyle('order: 3')
    expect(screen.getByText('Testing col 2')).toHaveStyle('order: 1')
    expect(screen.getByText('Testing col 3')).toHaveStyle('order: 2')
  })

  // test('should render correctly with breakpoints prop', () => {
  //   const breakpoints: ColProps['breakpoints'] = { mobile: { span: 12 }, tablet: { span: 6 } }
  //   render(
  //     <Row>
  //       <Col breakpoints={breakpoints} span={4}>
  //         Testing col 1
  //       </Col>
  //       <Col breakpoints={breakpoints} span={4}>
  //         Testing col 2
  //       </Col>
  //       <Col breakpoints={breakpoints} span={4}>
  //         Testing col 3
  //       </Col>
  //     </Row>
  //   )
  //
  //   expect(screen.getByTestId('Row')).toMatchSnapshot()
  //   expect(screen.getByText('Testing col 1')).toHaveClass('span-4')
  //   expect(screen.getByText('Testing col 2')).toHaveClass('span-4')
  //   expect(screen.getByText('Testing col 3')).toHaveClass('span-4')
  //
  //   /* eslint-disable-next-line no-global-assign */
  //   window = Object.assign(window, { innerWidth: 1000 })
  //   act(() => {
  //     window.dispatchEvent(new Event('resize'))
  //   })
  //   expect(screen.getByText('Testing col 1')).toHaveClass('span-6')
  //   expect(screen.getByText('Testing col 2')).toHaveClass('span-6')
  //   expect(screen.getByText('Testing col 3')).toHaveClass('span-6')
  //
  //   /* eslint-disable-next-line no-global-assign */
  //   window = Object.assign(window, { innerWidth: 500 })
  //   act(() => {
  //     window.dispatchEvent(new Event('resize'))
  //   })
  //   expect(screen.getByText('Testing col 1')).toHaveClass('span-12')
  //   expect(screen.getByText('Testing col 2')).toHaveClass('span-12')
  //   expect(screen.getByText('Testing col 3')).toHaveClass('span-12')
  // })
})
