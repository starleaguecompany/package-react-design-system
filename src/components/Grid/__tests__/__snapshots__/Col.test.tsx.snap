// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col should render correctly with attributes 1`] = `
<div
  class="test-class container"
  data-qa="Col"
  id="test-id"
>
  Testing col
</div>
`;

exports[`Col should render correctly with no props 1`] = `
<div
  class="container"
  data-qa="Col"
/>
`;

exports[`Col should render correctly with order prop 1`] = `
<div
  class="container wrap align-top justify-start"
  data-qa="Row"
>
  <div
    class="container"
    data-qa="Col"
    style="order: 3;"
  >
    Testing col 1
  </div>
  <div
    class="container"
    data-qa="Col"
    style="order: 1;"
  >
    Testing col 2
  </div>
  <div
    class="container"
    data-qa="Col"
    style="order: 2;"
  >
    Testing col 3
  </div>
</div>
`;

exports[`Col should render correctly with span prop 1`] = `
<div
  class="container span-6"
  data-qa="Col"
>
  Testing col
</div>
`;

exports[`Col should render correctly with text 1`] = `
<div
  class="container"
  data-qa="Col"
>
  Testing col
</div>
`;
